### PR TITLE
26 db migration, type safety, domain layer and insert project api

### PR DESCRIPTION
**Pull request by francisduvivier**
# Upload API PR

This PR is preparation for enabling app uploads on badgehub. For enabling this, a lot of stuff was changed as described below.  A lot of stuff still has to be done as well, but this PR is feature par with what we currently have.
## Related PR
https://github.com/badgeteam/badgehub-frontend/pull/16
### db-migrate

installed db-migrate and added the new table structure as a migration

### sql-tagged-template

all sql code is now using sql-tagged-templates which helps with readability, README.md was also updated for this

### Insert Project

You can now insert a project, which will also create an empty version and metadata and this project can then be retrieved

### Publish project

The publish patch api on a project will mark the latest version as published now

### Write API only enabled in development mode

Since we don't have security on our api yet, I added code that makes sure that the upload api only works when the NODE_ENV is set to development

### Migrate sample data

- Sample data was migrated by re-exporting the database with pg_dump

### Tested with frontend for feature parity compared to master

- with filtering on category and device working

### Consistent naming: 
- Is it apps or projects, is it devices or badges?

### Single File upload

You should be able to upload a file and this file should be put in the db and stored on disk using it's hash as the filename, like git does it, so also with 2-letter subdirs ideally.
if the file path is /metadata.json, then data should be extracted and filled in on the app_metadata_jsons table

### Zip upload

When you upload a zip, file upload should be done for all the files in the zip (so also extracting metadata)
**Q**: should zip upload also do publish?
**A**: no, it should not, there is an api for that. optionally, we can make it a query param later

### (one/many) to many relations from the [relations](src/db/models/app/relations) dir

- The one that is really needed for project upload is the file to version relation, others can wait.
- This will allow eg. finding all the collaborators on a project, but also all projects that a user has collaborated on...

### Maybe change http status codes and message

- currently, it will return 204 for the write api's, maybe that should be 200 OK?
- if you do a publish but it is already published, maybe that should be a different status code?

### Maybe improve SQL debugging

- We currently don't see the sql queries printed, even if they fail, maybe we should add that.

### Change it to make it clear that draft version is always unpublished

=> so publish should also create a new version

### CODE REVIEW

**Comment by paulinevos** [comment]
Does this need to be rebased or does it actually have 102 commits?

File: .env.example [hunk 1]
```diff
@@ -1,6 +1,7 @@
-POSTGRES_HOST=db
```
Comment (paulinevos) [inline .env.example:1]: Why change this? Would need to be `db` on the docker network, no?

**Comment by francisduvivier** [comment]
Note:
- I received a comment that this PR is too big for reviewing and I agree, so I am currently in the works of splitting it up.
- I found that the populateDB.ts script is not updated yet, I'll still do that.

**Review by zera1ul — APPROVED** [review]

File: .env.example (hunk 1 above)
Comment (francisduvivier) [reply .env.example]: check, changed back.

**Comment by francisduvivier** [comment]
I'm closing this PR and replacing it with a cleaner more split up version:

Part 1: #33 
Part 2: #35 
Part 3: #34

**Pull request closed without merging**